packages/flutter/lib/src/material/material_state.dart

```diff
@@ -184,3 +186,19 @@ class _MaterialStateColor extends MaterialStateColor {
   @override
   Color resolve(Set<MaterialState> states) => _resolve(states);
 }
+
+/// Interface for classes that can return a value of type `T` based on a set of
+/// [MaterialState]s.
+///
+/// For example, [MaterialStateColor] implements `MaterialStateProperty<Color>`
+/// because it has a `resolve` method that returns a different [Color] depending
+/// on the given set of [MaterialState]s.
+abstract class MaterialStateProperty<T> {
+
+  /// Returns a different value of type `T` depending on the given `states`.
+  ///
+  /// Some widgets (such as [RawMaterialButton]) keep track of their set of
+  /// [MaterialState]s, and will call `resolve` with the current states at build
+  /// time for specified properties (such as [RawMaterialButton.textStyle]'s color).
+  T resolve(Set<MaterialState> states);
```

Comment:
Why not include the generic function here as well:
```
  static T resolve<T>(T value, Set<MaterialState> states) {
    if (value is MaterialStateProperty<T>) {
      final MaterialStateProperty<T> property = value;
      return property.resolve(states);
    }
    return value;
  }
```

You could replace `MaterialState.resolveColor()` with `MaterialProperty.resolve<Color>()`.

Reply:
Great idea! Done.